accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,502 @@
+
+The proposed APIs will be used by the application or Library authors to record metrics measurements. The APIs also include a listener which will allow listening to the recorded measurements events which can be used by the OpenTelemetry SDK to aggregate the measurements and export them.
+
+The proposed APIs are intended to be supported on Full Framework 4.6 and up, .NET Core all versions, and .NET 5.0 and up.
+
+```MD040
+Naming in this proposal is picked up from the OpenTelemetry specification to avoid having .NET APIs
+deviating from the standard even if the standard names are not the best to use.
+```
+
+## Metric Terminologies
+
+### Instrument
+
+The Instrument is the type that will be used by the app and library authors to report measurements (e.g. Counter, Gauge...etc.). The instrument will have a name that should be validated as described in the OpenTelemetry specs. The Instrument can optionally have a description and unit of measurements. Instruments will be created with a numerical type parameter (e.g. `Counter<int>`). The Instrument is going to support all CLS-compliant numerical types which are supported on the Full Framework (Byte, Int16, Int32, Int64, Single, Double, and Decimal).
+
+There are two types of instruments:
+
+- The first type we'll call it just `Instrument` for simplicity. These instruments are called inside a request, meaning they have an associated distributed Context (with Span, Baggage, etc.). OpenTelemetry specs call this type of instrument a synchronous Instrument but we are trying to avoid confusion with the async feature of the .NET. The proposal here proposes three instrument classes of that type: `Counter`, `UpDownCounter`, and `Histogram`.
+- The second type is called `ObservableInstrument` which reports measurements by a callback, once per collection interval, and lacks Context. OpenTelemetry specs call this type of instrument an asynchronous Instrument but we are trying to avoid confusion with the async feature of the .NET. The proposal here is proposing three instrument classes of that type: `ObservableCounter`, `ObservableGauge`, and `ObservableUpDownCounter`.
+
+### Meter
+
+Meter is the factory type responsible for creating Instruments. Meter will have a name and optional version strings.
+
+### Listener
+
+The listener is the type that allows listening to the recorded measurements occurred by both the non-observable instrument and the observable instrument. The listener is an important class that will be used by OpenTelemetry to implement the Metrics SDK.
```

Comment:
The "non-observable instrument" sounds a bit weird 🧙

Reply:
agree. do you have a better name I can use here? I am trying to avoid using the `sync/async` terms.

Reply:
Perhaps "The listener is the type that allows listening to the measurements reported by the instruments (e.g. `Counter`, `ObservableCounter`, etc.)."